# Exploring Ebay Car Sales

In this project, I work with a dataset of used cars from [eBay Kleinanzeigen](https://www.ebay-kleinanzeigen.de/), a classifieds section of the German eBay website. The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database). I used a sample of 50,000 data points from the full dataset. 

I cleaned the column names, explored the odometer, price, and date columns, dealt with incorrect registration year data, explored price by brand, and stored aggregate data in a dataframe.

The aim of this project is to clean the data and analyze the included used car listings.

**The data dictionary provided with data is as follows:**

dateCrawled - When this ad was first crawled. All field-values are taken from this date.

name - Name of the car.

seller - Whether the seller is private or a dealer.

offerType - The type of listing

price - The price on the ad to sell the 
car.
abtest - Whether the listing is included in an A/B test.

vehicleType - The vehicle Type.

yearOfRegistration - The year in which the car was first registered.

gearbox - The transmission type.

powerPS - The power of the car in PS.

model - The car model name.

kilometer - How many kilometers the car has driven.

monthOfRegistration - The month in which the car was first registered.

fuelType - What type of fuel the car uses.

brand - The brand of the car.

notRepairedDamage - If the car has a damage which is not yet repaired.

dateCreated - The date on which the eBay listing was created.

nrOfPictures - The number of pictures in the ad.

postalCode - The postal code for the location of the vehicle.

lastSeenOnline - When the crawler saw this ad last online.

## Import

I start by importing the pandas library, which is good for manipulating and analyzing data. I use pandas to read the csv file into a pandas dataframe, storing it into a variable.

In [47]:
import pandas as pd

autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [48]:
print(autos.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [49]:
print(len(autos))
autos.head()

50000


dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [50]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


## Observations
The autos dataset includes 50,000 automobiles with 20 columns of information including their name, seller, offer type, price, role in an abtest, vehicle type, registration year, transmission type, power (in PS), model name, distance travelled (in kilometers), registration month, fuel type, brand, damage, number of pictures in the ad, zip code. 

** Note: PS stands for 'Pferdestärke' in German, which means horsepower **

1. Most of the columns are strings.

2. Some columns have null values, but none have more than ~20% null values.

3. The column names use camelcase rather than snakecase, which means that I can't just replace spaces with underscores. 

There is a study that shows that readers can recognize snake_case values more quickly than camelCase, so it is always better to use snake_case in Python. 

Also, some column names are not as descriptive as the data dictionary allows.

## Cleaning

I created a column name dictionary, which I used to reword some of the column names based on the data dictioanry, and I changed them from camelCase to snake_case by adding an underscore.

In [51]:
print(autos.columns)

column_name_dict = {
    'yearOfRegistration' : 'registration_year',
    'monthOfRegistration' : 'registration_month',
    'notRepairedDamage' : 'unrepaired_damage',
    'dateCreated' : 'ad_created'   ,
    'dateCrawled' : 'date_crawled',
    'offerType' : 'offer_type',
    'vehicleType' : 'vehicle_type',
    'powerPS' : 'power_ps',
    'fuelType' : 'fuel_type',
    'nrOfPictures' : 'nr_of_pictures',
    'postalCode' : 'postal_code',
    'lastSeen' : 'last_seen'
}

autos.rename(columns=column_name_dict, inplace=True)


Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [52]:
print(autos.columns)
autos.head()

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [53]:
autos.describe(include='all') #both categorial and numeric columns

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-30 19:48:02  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

I will note:

-Any columns that have mostly one value are condidates to be dropped
        
        ex. seller, offer_type (Angebot and privat are the most common, and occur 49,999 times)
        
-Columns that need more investigation
    
-Numeric data stored as text that needs to be cleaned

        ex. price and odometer

For the price and odometer columns, I will remove non-numeric characters, convert the column to a numeric dtype, and rename the column.


In [54]:
autos['price'] = autos['price'].str.replace('$', '')
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['price'] = autos['price'].str.replace(',', '')
autos['odometer'] = autos['odometer'].str.replace(',','')
autos.rename(columns={'odometer' : 'odometer_km'}, inplace=True)

In [55]:
autos['price'] = autos['price'].astype(float)
autos['odometer_km'] = autos['odometer_km'].astype(float)

type(autos['price'])          #series
autos['price'].dtype          #float64
type(autos['odometer_km'])    #series
autos['odometer_km'].dtype    #float64

dtype('float64')

In [56]:
autos['price'].unique()

array([ 5000.,  8500.,  8990., ...,   385., 22200., 16995.])

In [57]:
autos['price'].unique().shape

(2357,)

In [58]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [59]:
autos['price'].value_counts()
autos['price'].value_counts(ascending=True)
autos['price'].value_counts(ascending=False)
autos['price'].value_counts(ascending=False)
autos['price'].value_counts().sort_index(ascending=True).tail()

10000000.0    1
11111111.0    2
12345678.0    3
27322222.0    1
99999999.0    1
Name: price, dtype: int64

In [60]:
autos = autos[autos['price'].between(0,30000)]

In [61]:
autos['odometer_km'].unique()

array([150000.,  70000.,  50000.,  80000.,  10000., 125000.,  90000.,
        20000.,  60000.,   5000.,  30000., 100000.,  40000.])

In [62]:
autos['odometer_km'].unique().shape

(13,)

In [63]:
autos['odometer_km'].describe()

count     49206.000000
mean     126770.210950
std       39127.884921
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [64]:
autos['odometer_km'].value_counts(ascending=True)

10000.0       217
20000.0       681
30000.0       711
40000.0       747
5000.0        921
50000.0       949
60000.0      1105
70000.0      1178
80000.0      1393
90000.0      1715
100000.0     2120
125000.0     5124
150000.0    32345
Name: odometer_km, dtype: int64

In [65]:
autos['odometer_km'].value_counts().sort_index(ascending=True)

5000.0        921
10000.0       217
20000.0       681
30000.0       711
40000.0       747
50000.0       949
60000.0      1105
70000.0      1178
80000.0      1393
90000.0      1715
100000.0     2120
125000.0     5124
150000.0    32345
Name: odometer_km, dtype: int64

In [67]:
autos = autos[autos['odometer_km'].between(50000, 150000)]

**Outliers**

I removed the prices below $30,000, because they occur less frequently.

Most cars are priced under $30,000.

I removed the odometer values below 50,000km because they occur less frequently. Most cars have travelled 150,000 km. Only 25% of the cars travelled 125,000km or below.

The average price is $9,840. There are a lot of free cars.
The average distance travelled is 126,770km.

In [68]:
#managing the dates columns

autos['date_crawled'].dtype        #object
autos['last_seen'].dtype           #object
autos['ad_created'].dtype          #object
autos['registration_month'].dtype  #int64
autos['registration_year'].dtype   #int64

autos['registration_month'].describe()

count    45929.000000
mean         5.756211
std          3.711413
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: registration_month, dtype: float64

In [69]:
autos['registration_year'].describe()

count    45929.000000
mean      2003.223671
std         33.412557
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2007.000000
max       9000.000000
Name: registration_year, dtype: float64

In [70]:
#converting string columns into dates to understand quantitatively

autos[['date_crawled', 'ad_created', 'last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [71]:
#Shorten the date to only include the time
autos['date_crawled'].str[:10].head()
autos['ad_created'].str[:10].head()
autos['last_seen'].str[:10].head()

#Calculate the distribution of the dates as percentages, including missing values
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False)
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False)
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False)

#Rank the date in ascending order
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001045
2016-03-06    0.004572
2016-03-07    0.005596
2016-03-08    0.007969
2016-03-09    0.010124
2016-03-10    0.010908
2016-03-11    0.013042
2016-03-12    0.023950
2016-03-13    0.009123
2016-03-14    0.012998
2016-03-15    0.016025
2016-03-16    0.016656
2016-03-17    0.028609
2016-03-18    0.007403
2016-03-19    0.016242
2016-03-20    0.021316
2016-03-21    0.021207
2016-03-22    0.021838
2016-03-23    0.018746
2016-03-24    0.020074
2016-03-25    0.019552
2016-03-26    0.017048
2016-03-27    0.016330
2016-03-28    0.021381
2016-03-29    0.022818
2016-03-30    0.025431
2016-03-31    0.024190
2016-04-01    0.023645
2016-04-02    0.024952
2016-04-03    0.025539
2016-04-04    0.024995
2016-04-05    0.121753
2016-04-06    0.217836
2016-04-07    0.127087
Name: last_seen, dtype: float64

Observations
The date crawled picked up on March 7th.
The date crawled dropped after April 4th. 
The site was crawled from March 5th, 2016 to April 7th, 2016.

The ads were created between August 10th, 2015 to July, 4th 2016. The ads created in late 2015 and the beginning of 2016 were significantly lower than the ads created in the Spring. This could mean that March is a more popular month for selling cars.

The last-seen dates range between March, 3rd, 2016 and April 7th, 2016. April 5th, 2016 and April 6th, 2016 are the most frequent last seen dates. There are a greater number of last seen dates in mid-March and early-April.

In [72]:
autos[['registration_month', 'registration_year']].head()

registration_month  registration_year
0                   3               2004
1                   6               1997
2                   7               2009
3                   6               2007
4                   7               2003

In [73]:
autos[['registration_month', 'registration_year']].describe()

registration_month  registration_year
count        45929.000000       45929.000000
mean             5.756211        2003.223671
std              3.711413          33.412557
min              0.000000        1910.000000
25%              3.000000        1999.000000
50%              6.000000        2003.000000
75%              9.000000        2007.000000
max             12.000000        9000.000000

Understanding the distribution of registration_year

The cars' registration year range from 1910 to 9000. 
Most of the cars were registered around 2003. 
The oldest 25% of the cars were registered on or before 1999. 
The newest 25% of the cars were registered on or after 2007.

**9000 could have been incorrectly entered on the website
The highest acceptable value is 2016 because a car can't be first registered after the listing was seen. 

In [74]:
#Removing values outside the upper and lower bounds
#Calculating the distribution of the remaining values
autos['registration_year'].value_counts(normalize=True).sort_index()

autos = autos[autos['registration_year'].between(0,2016)]

autos['registration_year'].value_counts(normalize=True).sort_index().tail()

autos[['registration_month', 'registration_year']].describe()

registration_month  registration_year
count        44092.000000       44092.000000
mean             5.800848        2002.480722
std              3.696457           6.621160
min              0.000000        1910.000000
25%              3.000000        1999.000000
50%              6.000000        2003.000000
75%              9.000000        2007.000000
max             12.000000        2016.000000

After removing the unrealistic "9000" registration year. The latest year that a car was registered in this dataset is 2016.

**Analysis technique: explore variations across different categories

When working with data on cars, it's natural to explore variations across different car brands. Aggregation can help with understanding the brand column.

In [75]:
#Exploring the unique values in the brand column, to decide on which brands I want to aggregate by.

autos[['brand']]
autos['brand'].unique()
autos['brand'].value_counts(normalize=True) > .05
autos['brand'].value_counts(normalize=True)[:20]
autos['brand'].value_counts()[:20]
autos['brand'].value_counts()[:20].index

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'mazda', 'skoda', 'nissan', 'citroen',
       'smart', 'toyota', 'volvo', 'hyundai', 'mitsubishi', 'honda'],
      dtype='object')

In [76]:
brand_index = (autos['brand'].value_counts(normalize=True) > .05).index[:6]
brand_index.dtype
brand_index

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

The top 20 are Volkswagen, BMW, Opel, Mercedez-Benz, Audi, and Ford,
Renault, Peugeot, Fiat, Seat, Mazda, Skoda, Nissan, Citroen, Smart, Toyota, Volvo, Hyundai, Mitsubishi, and Honda.

The top 5% are Volkswagen, BMW, Opel, Mercedez-Benz, Audi, and Ford.

I've chosen to aggregate on the top 5% of car brands. 

In [77]:
#Creating a dictionary to hold the aggregate data
#Calculating the mean mileage and mean price for each of the top brands

mean_price = {}
mean_mileage = {}

for brand in brand_index:
    mean_price[brand] = autos.groupby(['brand'])['price'].mean()[brand]
    mean_mileage[brand] = autos.groupby(['brand'])['odometer_km'].mean()[brand]

In [78]:
mean_price

{'audi': 7219.228275312251,
 'bmw': 7068.200040371417,
 'ford': 3000.413449163449,
 'mercedes_benz': 6984.533050047215,
 'opel': 2595.4753629114703,
 'volkswagen': 4580.633301657692}

In [79]:
mean_mileage

{'audi': 138874.56816369918,
 'bmw': 138901.89745660074,
 'ford': 131957.8507078507,
 'mercedes_benz': 138059.49008498582,
 'opel': 136048.86526272746,
 'volkswagen': 136550.52264808363}

In [80]:
#Converting both dictionaries to series objects with the series constructor

import pandas as pd
mean_price_series = pd.Series(mean_price)
mean_mileage_series = pd.Series(mean_mileage)

mean_price_series
mean_mileage_series

audi             138874.568164
bmw              138901.897457
ford             131957.850708
mercedes_benz    138059.490085
opel             136048.865263
volkswagen       136550.522648
dtype: float64

In [81]:
#Creating a dataframe from the first series object w/ the df constructor

mean_price_df = pd.DataFrame(data = mean_price_series, columns=['mean_price'])
mean_price_df
mean_price_df['mean_mileage'] = mean_mileage_series
mean_price_df

mean_price   mean_mileage
audi           7219.228275  138874.568164
bmw            7068.200040  138901.897457
ford           3000.413449  131957.850708
mercedes_benz  6984.533050  138059.490085
opel           2595.475363  136048.865263
volkswagen     4580.633302  136550.522648

In [82]:
#pretty printing the dataframe

with pd.option_context('display.max_rows', None, 'display.max_columns', None):  # more options can be specified also
    print(mean_price_df)

                mean_price   mean_mileage
audi           7219.228275  138874.568164
bmw            7068.200040  138901.897457
ford           3000.413449  131957.850708
mercedes_benz  6984.533050  138059.490085
opel           2595.475363  136048.865263
volkswagen     4580.633302  136550.522648


# Analysis

In this project, I applied a variety of pandas methods to explore and understand a dataset on car listings.

In the top 6 brands, there's a distinct price gap. Audi, BMW, and Mercedez Benz are most expenseive, Ford and Opel are less expensive. Volkswagen is in between. 

For the top 6 brands, I used aggregation to understand their average mileage and if there was any link with mean price. I combined the data from both series objects into a single dataframe (with a shared index) and displayed the dataframe directly.

I could have displayed both the aggregated series objects and visually compared them, however that would be inconvenient for extending this to more columns. When comparing aggregate series objects, it is better to combine the data from both series objects into a single dataframe.

There seems to be a link between the mean price and the mean mileage. The more expensie cars have a higher mean mileage among the top 6 brands. 

** Here are some next steps that I could take: **

-Identify categorical data that uses german words, translate them and map the values to their english couterparts

-Convert the dates to be uniform numeric data.

-See if there are particular words in the name column that I can extract as new columns

** For further analysis, I could: **

-Find the most common brand/model combinations

-Split odometer_km into groups, and use aggregation to see if average
proces follows any patterns based on mileage

-Answer the question "How much cheaper are cars with damage than their non-damaged counterparts?"